File: backend/services/cache.service.ts
/**
 * 24-hour In-Memory Assessment Cache with TTL
 */
import type { AssessResponse } from '../models/types.js';

interface CacheEntry {
  data: AssessResponse;
  timestamp: number;
}

const TTL_MS = 24 * 60 * 60 * 1000; // 24 hours
const cache = new Map<string, CacheEntry>();

export function getCacheKey(lat: number, lon: number): string {
  return `${lat.toFixed(3)},${lon.toFixed(3)}`;
}

export function getCachedAssessment(lat: number, lon: number): AssessResponse | null {
  const key = getCacheKey(lat, lon);
  const entry = cache.get(key);
  if (!entry) return null;

  if (Date.now() - entry.timestamp > TTL_MS) {
    cache.delete(key);
    return null;
  }

  return entry.data;
}

export function setCachedAssessment(lat: number, lon: number, data: AssessResponse): void {
  const key = getCacheKey(lat, lon);
  cache.set(key, {
    data,
    timestamp: Date.now(),
  });
}

File: backend/services/cache_service.py
"""
24-hour In-Memory Assessment Cache
===================================
Direct port of `backend/services/cache.service.ts`.
"""

from __future__ import annotations

import time
from typing import Any


class AssessmentCache:
    """In-memory cache for risk assessment results with 24h TTL."""

    _TTL_MS: int = 24 * 60 * 60 * 1000  # 24 hours
    _cache: dict[str, tuple[float, Any]] = {}  # key -> (timestamp, data)

    @staticmethod
    def _make_key(lat: float, lon: float) -> str:
        return f"{lat:.3f},{lon:.3f}"

    @staticmethod
    def get(lat: float, lon: float) -> Any | None:
        key = AssessmentCache._make_key(lat, lon)
        entry = AssessmentCache._cache.get(key)
        if entry is None:
            return None
        timestamp, data = entry
        if (time.time() * 1000 - timestamp) > AssessmentCache._TTL_MS:
            del AssessmentCache._cache[key]
            return None
        return data

    @staticmethod
    def set(lat: float, lon: float, data: Any) -> None:
        key = AssessmentCache._make_key(lat, lon)
        AssessmentCache._cache[key] = (time.time() * 1000, data)
